crates/red_knot_test/src/diagnostic.rs

```diff
@@ -0,0 +1,125 @@
+//! Sort and group diagnostics by line number, so they can be correlated with assertions.
+//!
+//! We don't assume that we will get the diagnostics in source order.
+use ruff_source_file::{LineIndex, OneIndexed};
+use ruff_text_size::Ranged;
+use smallvec::SmallVec;
+
+#[derive(Debug)]
+pub(crate) struct SortedDiagnostics<T>(Vec<DiagnosticWithLine<T>>);
+
+impl<T> SortedDiagnostics<T>
+where
+    T: Ranged + Clone,
+{
+    pub(crate) fn new(diagnostics: impl IntoIterator<Item = T>, line_index: &LineIndex) -> Self {
+        let mut diagnostics = diagnostics
+            .into_iter()
+            .map(|diagnostic| DiagnosticWithLine {
+                line: line_index.line_index(diagnostic.range().start()),
+                diagnostic,
+            })
+            .collect::<Vec<_>>();
+        diagnostics.sort_by_key(|diag| diag.line);
+
+        Self(diagnostics)
+    }
+
+    pub(crate) fn iter_lines(&self) -> LineDiagnosticsIterator<T> {
+        LineDiagnosticsIterator {
+            inner: self.0.iter().peekable(),
+        }
+    }
+}
+
+/// Iterator to group sorted diagnostics by line.
+pub(crate) struct LineDiagnosticsIterator<'a, T> {
+    inner: std::iter::Peekable<std::slice::Iter<'a, DiagnosticWithLine<T>>>,
+}
+
+impl<T> Iterator for LineDiagnosticsIterator<'_, T>
+where
+    T: Ranged + Clone,
+{
+    type Item = LineDiagnostics<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let diag = self.inner.next()?;
+        let line = diag.line;
+        let mut diagnostics = DiagnosticVec::new();
+        diagnostics.push(diag.diagnostic.clone());
```

Comment:
I tried this and found the lifetimes difficult to manage. In practice (at least with the current implementation of `TypeCheckDiagnostics`) the concrete type of `T` here is `Arc<TypeCheckDiagnostic>`, so we're just cloning an `Arc`, which isn't too expensive. If that changes in a future re-work of diagnostics and cloning gets more costly, then it might be worth revisiting this to try with references again.